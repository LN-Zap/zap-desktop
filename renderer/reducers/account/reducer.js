import createReducer from '@zap/utils/createReducer'
import { getIntl } from '@zap/i18n'
import { mainLog } from '@zap/utils/log'
import waitForIpcEvent from '@zap/utils/waitForIpc'
import { closeDialog } from 'reducers/modal'
import { showNotification } from 'reducers/notification'
import accountSelectors from './selectors'
import messages from './messages'
import * as constants from './constants'

const {
  INIT_ACCOUNT,
  INIT_ACCOUNT_SUCCESS,
  INIT_ACCOUNT_FAILURE,
  LOGIN,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGIN_CLEAR_ERROR,
  CHANGE_PASSWORD_DIALOG_ID,
  PASSWORD_PROMPT_DIALOG_ID,
  PASSWORD_SET_DIALOG_ID,
  SET_IS_PASSWORD_ENABLED,
  LOGIN_NOT_ALLOWED,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

export const initialState = {
  isAccountLoading: false,
  isAccountLoaded: false,
  initAccountError: null,
  isLoggingIn: false,
  loginError: null,
  isLoggedIn: false,
  isPasswordEnabled: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * initAccount - Fetch the current account info from the database and save into the store.
 * Should be called once when the app first loads.
 *
 * @returns {Function} Thunk
 */
export const initAccount = () => async dispatch => {
  dispatch({ type: INIT_ACCOUNT })
  try {
    // Auto login user if password feature is disabled.
    const isAccountPasswordEnabled = await dispatch(checkAccountPasswordEnabled())
    if (!isAccountPasswordEnabled) {
      dispatch({ type: LOGIN_SUCCESS })
    }
    dispatch({ type: INIT_ACCOUNT_SUCCESS })
  } catch (e) {
    mainLog.warn('checkAccountPasswordEnabled error: %o', e)
    dispatch({ type: LOGIN_FAILURE, error: LOGIN_NOT_ALLOWED })
  }
}

/**
 * setIsPasswordEnabled - Whether password is set for the account.
 *
 * @param {string} value value
 * @returns {object} Action
 */
const setIsPasswordEnabled = value => ({
  type: SET_IS_PASSWORD_ENABLED,
  value,
})

/**
 * setPassword - Updates wallet password.
 *
 * @param {string} password new password
 * @returns {Function} Thunk
 */
const setPassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  dispatch(waitForIpcEvent('setPassword', { value: await sha256digest(password) }))
}

/**
 * checkAccountPasswordEnabled - Checks whether app password is set via checking secure storage.
 * Dispatches setIsPasswordEnabled on completion.
 *
 * @returns {Function} Thunk
 */
const checkAccountPasswordEnabled = () => async dispatch => {
  const { value } = await dispatch(waitForIpcEvent('hasPassword'))
  dispatch(setIsPasswordEnabled(value))
  return value
}

/**
 * changePassword - Changes existing password.
 *
 * @param {object} params password params
 * @param {string} params.newPassword new password
 * @param {string} params.oldPassword old password
 * @returns {Function} Thunk
 */
export const changePassword = ({ newPassword, oldPassword }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(oldPassword))
    await dispatch(setPassword(newPassword))
    dispatch(closeDialog(CHANGE_PASSWORD_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_updated)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * enablePassword - Enables app-wide password protection.
 *
 * @param {string} password to be used further on.
 * @returns {Function} Thunk
 */
export const enablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(setPassword(password))
    dispatch(closeDialog(PASSWORD_SET_DIALOG_ID))
    dispatch(setIsPasswordEnabled(true))
    dispatch(showNotification(intl.formatMessage(messages.account_password_enabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * disablePassword - Disables app-wide password protection.
 *
 * @param {string} password current password.
 * @returns {Function} Thunk
 */
export const disablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(password))
    await dispatch(waitForIpcEvent('deletePassword'))
    dispatch(setIsPasswordEnabled(false))
    dispatch(closeDialog(PASSWORD_PROMPT_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_disabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * requirePassword - Password protect routine. Should be placed before protected code.
 *
 * @param {string} password current password.
 * @returns {Promise} Promise that fulfills after login attempt (either successful or not)
 */
const requirePassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  const { password: hash } = await dispatch(waitForIpcEvent('getPassword'))
  const passwordHash = await sha256digest(password)
  // compare hash received from the main thread to a hash of a password provided
  if (hash === passwordHash) {
    return true
  }
  throw new Error(getIntl().formatMessage(messages.account_invalid_password))
}

/**
 * login - Perform account login.
 *
 * @param {string} password Password
 * @returns {Function} Thunk
 */
export const login = password => async dispatch => {
  try {
    dispatch({ type: LOGIN, password })
    await dispatch(requirePassword(password))
    dispatch({ type: LOGIN_SUCCESS })
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * clearLoginError - Clear login user error.
 *
 * @returns {Function} Thunk
 */
export const clearLoginError = () => (dispatch, getState) => {
  if (accountSelectors.loginError(getState())) {
    dispatch({
      type: 'LOGIN_CLEAR_ERROR',
    })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [INIT_ACCOUNT]: state => {
    state.isAccountLoading = true
  },
  [INIT_ACCOUNT_SUCCESS]: state => {
    state.isAccountLoading = false
    state.isAccountLoaded = true
    state.initAccountError = null
  },
  [INIT_ACCOUNT_FAILURE]: (state, { error }) => {
    state.isAccountLoading = false
    state.initAccountError = error
  },

  [LOGIN]: state => {
    state.isLoggingIn = true
  },
  [LOGIN_SUCCESS]: state => {
    state.isLoggingIn = false
    state.isLoggedIn = true
    state.loginError = null
  },
  [LOGIN_FAILURE]: (state, { error }) => {
    state.isLoggingIn = false
    state.loginError = error
  },
  [LOGIN_CLEAR_ERROR]: state => {
    state.loginError = null
  },
  [SET_IS_PASSWORD_ENABLED]: (state, { value }) => {
    state.isPasswordEnabled = value
  },
}

// ------------------------------------
// Reducer
// ------------------------------------

export default createReducer(initialState, ACTION_HANDLERS)
