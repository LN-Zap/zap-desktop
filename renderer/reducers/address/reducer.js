import get from 'lodash/get'

import { getIntl } from '@zap/i18n'
import createReducer from '@zap/utils/createReducer'
import { openModal, closeModal } from 'reducers/modal'
import { showError } from 'reducers/notification'
import { settingsSelectors } from 'reducers/settings'
import { grpc } from 'workers'

import messages from '../messages'
import * as constants from './constants'

const {
  FETCH_ADDRESSES,
  FETCH_ADDRESSES_SUCCESS,
  FETCH_ADDRESSES_FAILURE,
  NEW_ADDRESS,
  NEW_ADDRESS_SUCCESS,
  NEW_ADDRESS_FAILURE,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isAddressLoading Boolean indicating if addresses are loading.
 * @property {Error|null} fetchAddressError Error from loading addresses.
 * @property {{np2wkh: boolean, p2wkh: boolean}} addressesLoading Booleans indicating address types that are loading.
 * @property {{np2wkh: string|null, p2wkh: string|null}} addresses Current bitcoin addresses.
 */

/** @type {State} */
const initialState = {
  isAddressLoading: false,
  fetchAddressError: null,
  addressesLoading: {
    np2wkh: false,
    p2wkh: false,
  },
  addresses: {
    np2wkh: null,
    p2wkh: null,
  },
}

// LND expects types to be sent as int, so this object will allow mapping from string to int
const ADDRESS_TYPES = {
  p2wkh: 0,
  np2wkh: 1,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * openWalletModal - Open wallet modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const openWalletModal = () => dispatch => dispatch(openModal('RECEIVE_MODAL'))

/**
 * closeWalletModal - Close wallet modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const closeWalletModal = () => dispatch => dispatch(closeModal('RECEIVE_MODAL'))

/**
 * newAddressSuccess - Generate new address success callback.
 *
 * @param {string} addressType Address type
 * @param {string} address Address
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const newAddressSuccess = (addressType, address) => async (dispatch, getState) => {
  const state = getState()
  const pubKey = state.info.data.identityPubkey

  // If we know the node's public key, store the address for reuse.
  if (pubKey) {
    const node = await window.db.nodes.get(pubKey)
    if (node) {
      node.setCurrentAddress(addressType, address)
      await window.db.nodes.put(node)
    } else {
      await window.db.nodes.put({ id: pubKey, addresses: { [addressType]: address } })
    }
  }

  dispatch({ type: NEW_ADDRESS_SUCCESS, addressType, address })
}

/**
 * createNewAddress - Creates a new address.
 *
 * @param {('p2wkh'|'np2wkh'|null)} addressType Address type. If not specified uses current saved config setting.
 * @returns {(dispatch:Function, getState:Function) => Promise<string>} Thunk
 */
export const createNewAddress = addressType => async (dispatch, getState) => {
  const getConfigAddressType = () => {
    const settings = settingsSelectors.currentConfig(getState())
    return settings && settings.address
  }
  const type = addressType || getConfigAddressType()
  const data = await grpc.services.Lightning.newAddress({ type: ADDRESS_TYPES[type] })
  return data.address
}

/**
 * newAddressFailure - Generate new address failure callback.
 *
 * @param {string} addressType Address type
 * @param {string} error Error message
 * @returns {(dispatch:Function) => void} Thunk
 */
export const newAddressFailure = (addressType, error) => dispatch => {
  const intl = getIntl()
  dispatch(
    showError(intl.formatMessage(messages.address_new_address_error, { addressType, error }))
  )
  dispatch({
    type: NEW_ADDRESS_FAILURE,
    addressType,
    error,
  })
}

/**
 * newAddress - Generate a new address and sets it as wallet current.
 *
 * @param {('p2wkh'|'np2wkh')} addressType Address type
 * @returns {(dispatch:Function) => Promise<string>} Thunk
 */
export const newAddress = addressType => async dispatch => {
  dispatch({ type: NEW_ADDRESS, addressType })
  try {
    const address = await dispatch(createNewAddress(addressType))
    await dispatch(newAddressSuccess(addressType, address))
    return address
  } catch (error) {
    dispatch(newAddressFailure(addressType, error.message))
    return null
  }
}

/**
 * initAddresses - Initialise addresses.
 *
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const initAddresses = () => async (dispatch, getState) => {
  try {
    dispatch({ type: FETCH_ADDRESSES })

    const state = getState()

    // Get node information (addresses are keyed under the node pubkey).
    const pubKey = state.info.data.identityPubkey
    const node = await window.db.nodes.get({ id: pubKey })

    // Get existing addresses for the node.
    const addresses = get(node, 'addresses', {})

    // Ensure that we have an address for all supported address types.
    const fetchAddresses = async types => {
      const data = {}
      await Promise.all(
        types.map(async type => {
          data[type] = addresses[type] || (await dispatch(newAddress(type)))
        })
      )
      return data
    }

    const allAddresses = await fetchAddresses(Object.keys(ADDRESS_TYPES))

    dispatch({ type: FETCH_ADDRESSES_SUCCESS, addresses: allAddresses })
  } catch (error) {
    dispatch({ type: FETCH_ADDRESSES_FAILURE, error: error.message })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_ADDRESSES]: state => {
    state.isAddressLoading = true
  },
  [FETCH_ADDRESSES_SUCCESS]: (state, { addresses }) => {
    state.addresses = addresses
    state.isAddressLoading = false
    state.fetchAddressError = null
  },
  [FETCH_ADDRESSES_FAILURE]: (state, { error }) => {
    state.isAddressLoading = false
    state.fetchAddressError = error
  },
  [NEW_ADDRESS]: (state, { addressType }) => {
    state.addressesLoading[addressType] = true
  },
  [NEW_ADDRESS_SUCCESS]: (state, { addressType, address }) => {
    state.addressesLoading[addressType] = false
    state.addresses[addressType] = address
  },
  [NEW_ADDRESS_FAILURE]: (state, { addressType }) => {
    state.addressesLoading[addressType] = false
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
