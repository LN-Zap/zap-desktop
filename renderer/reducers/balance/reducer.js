import Coin from '@zap/utils/coin'
import createReducer from '@zap/utils/createReducer'
import { grpc } from 'workers'

import * as constants from './constants'

const { FETCH_BALANCE, FETCH_BALANCE_SUCCESS, FETCH_BALANCE_FAILURE } = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isBalanceLoading Boolean indicating if balances are loading.
 * @property {string|null} walletBalance Total wallet balance.
 * @property {string|null} walletBalanceConfirmed Confirmed wallet balance.
 * @property {string|null} walletBalanceUnconfirmed Unconfirmed wallet balance.
 * @property {string|null} channelBalance Total channel balance.
 * @property {string|null} channelBalanceConfirmed Confirmed channel balance.
 * @property {string|null} channelBalancePending Pending channel balance.
 * @property {Error|null} fetchBalanceError Balance fetching error.
 */

/** @type {State} */
const initialState = {
  isBalanceLoading: false,
  walletBalance: null,
  walletBalanceConfirmed: null,
  walletBalanceUnconfirmed: null,
  channelBalance: null,
  channelBalanceConfirmed: null,
  channelBalancePending: null,
  fetchBalanceError: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * fetchBalance - Fetch balances.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const fetchBalance = () => async dispatch => {
  try {
    dispatch({ type: FETCH_BALANCE })
    const { walletBalance, channelBalance } = await grpc.services.Lightning.getBalance()
    dispatch({ type: FETCH_BALANCE_SUCCESS, walletBalance, channelBalance })
  } catch (error) {
    dispatch({ type: FETCH_BALANCE_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_BALANCE]: state => {
    state.isBalanceLoading = true
  },
  [FETCH_BALANCE_SUCCESS]: (state, { walletBalance, channelBalance }) => {
    state.isBalanceLoading = false
    state.walletBalance = walletBalance.totalBalance
    state.walletBalanceConfirmed = walletBalance.confirmedBalance
    state.walletBalanceUnconfirmed = walletBalance.unconfirmedBalance
    state.channelBalance = Coin(channelBalance.balance)
      .add(Coin(channelBalance.pendingOpenBalance))
      .toString()
    state.channelBalanceConfirmed = channelBalance.balance
    state.channelBalancePending = channelBalance.pendingOpenBalance
  },
  [FETCH_BALANCE_FAILURE]: (state, { error }) => {
    state.isBalanceLoading = false
    state.fetchBalanceError = error
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
