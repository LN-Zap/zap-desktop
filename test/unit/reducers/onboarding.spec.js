import reducer, {
  SET_CONNECTION_TYPE,
  SET_CONNECTION_STRING,
  SET_CONNECTION_HOST,
  SET_CONNECTION_CERT,
  SET_CONNECTION_MACAROON,
  SET_ALIAS,
  SET_NAME,
  SET_AUTOPILOT,
  SET_CHAIN,
  SET_NETWORK,
  SET_PASSWORD,
  SET_PASSPHRASE,
  SET_SEED,
  VALIDATING_HOST,
  VALIDATING_CERT,
  VALIDATING_MACAROON,
  RESET_ONBOARDING,
  SET_LNDCONNECT,
} from 'reducers/onboarding'

import snapshotDiff from '../__helpers__/snapshotDiff'

describe('reducers', () => {
  describe('onboardingReducer', () => {
    it('should handle initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle SET_CONNECTION_TYPE', () => {
      const action = {
        type: SET_CONNECTION_TYPE,
        connectionType: 'custom',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_CONNECTION_STRING', () => {
      const action = {
        type: SET_CONNECTION_STRING,
        connectionUri: 'lndconnect://localhost:10009?cert=/path/to/cert&macaroon=/path/to/macaroon',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_CONNECTION_HOST', () => {
      const action = {
        type: SET_CONNECTION_HOST,
        connectionHost: 'localhost',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_CONNECTION_CERT', () => {
      const action = {
        type: SET_CONNECTION_CERT,
        connectionCert: '/path/to/cert',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_CONNECTION_MACAROON', () => {
      const action = {
        type: SET_CONNECTION_MACAROON,
        connectionMacaroon: '/path/to/macaroon',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_ALIAS', () => {
      const action = {
        type: SET_ALIAS,
        alias: 'some alias',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_NAME', () => {
      const action = {
        type: SET_NAME,
        name: 'some name',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_AUTOPILOT', () => {
      const action = {
        type: SET_AUTOPILOT,
        autopilot: true,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_CHAIN', () => {
      const action = {
        type: SET_CHAIN,
        chain: 'bitcoin',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_NETWORK', () => {
      const action = {
        type: SET_NETWORK,
        network: 'testnet',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_PASSWORD', () => {
      const action = {
        type: SET_PASSWORD,
        password: 'some password',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_PASSPHRASE', () => {
      const action = {
        type: SET_PASSPHRASE,
        passphrase: 'some passphrase',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_SEED', () => {
      const action = {
        type: SET_SEED,
        seed: ['some', 'seed', 'words'],
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle VALIDATING_HOST', () => {
      const action = {
        type: VALIDATING_HOST,
        validatingHost: true,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle VALIDATING_CERT', () => {
      const action = {
        type: VALIDATING_CERT,
        validatingCert: true,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle VALIDATING_MACAROON', () => {
      const action = {
        type: VALIDATING_MACAROON,
        validatingMacaroon: true,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle RESET_ONBOARDING', () => {
      const action = {
        type: RESET_ONBOARDING,
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })

    it('should handle SET_LNDCONNECT', () => {
      const action = {
        type: SET_LNDCONNECT,
        lndConnect: 'lndconnect://localhost:10009?cert=/path/to/cert&macaroon=/path/to/macaroon',
      }
      expect(snapshotDiff(reducer(undefined, {}), reducer(undefined, action))).toMatchSnapshot()
    })
  })
})
