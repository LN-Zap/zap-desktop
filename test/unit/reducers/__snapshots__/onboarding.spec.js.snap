// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers onboardingReducer should handle RESET_ONBOARDING 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`reducers onboardingReducer should handle SET_ALIAS 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
-   "alias": "",
+   "alias": "some alias",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_AUTOPILOT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
-   "autopilot": false,
+   "autopilot": true,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_CHAIN 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`reducers onboardingReducer should handle SET_CONNECTION_CERT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
-   "connectionCert": "",
+   "connectionCert": "/path/to/cert",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_CONNECTION_HOST 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
-   "connectionHost": "",
+   "connectionHost": "localhost",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_CONNECTION_MACAROON 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
-   "connectionMacaroon": "",
+   "connectionMacaroon": "/path/to/macaroon",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_CONNECTION_STRING 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
-   "connectionString": "",
+   "connectionString": undefined,
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_CONNECTION_TYPE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
-   "connectionType": "create",
+   "connectionType": "custom",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_LNDCONNECT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
-   "lndConnect": null,
+   "lndConnect": "lndconnect://localhost:10009?cert=/path/to/cert&macaroon=/path/to/macaroon",
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_NAME 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
-   "name": "",
+   "name": "some name",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_NETWORK 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
-   "network": "mainnet",
+   "network": "testnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_PASSPHRASE 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
-   "passphrase": "",
+   "passphrase": "some passphrase",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_PASSWORD 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
-   "password": "",
+   "password": "some password",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle SET_SEED 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
-   "seed": Array [],
+   "seed": Array [
+     "some",
+     "seed",
+     "words",
+   ],
    "validatingCert": false,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle VALIDATING_CERT 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
-   "validatingCert": false,
+   "validatingCert": true,
    "validatingHost": false,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle VALIDATING_HOST 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
-   "validatingHost": false,
+   "validatingHost": true,
    "validatingMacaroon": false,
  }
`;

exports[`reducers onboardingReducer should handle VALIDATING_MACAROON 1`] = `
Snapshot Diff:
- First value
+ Second value

  Object {
    "alias": "",
    "autopilot": false,
    "chain": "bitcoin",
    "connectionCert": "",
    "connectionHost": "",
    "connectionMacaroon": "",
    "connectionString": "",
    "connectionType": "create",
    "isSkipBackupDialogOpen": false,
    "lndConnect": null,
    "name": "",
    "network": "mainnet",
    "passphrase": "",
    "password": "",
    "seed": Array [],
    "validatingCert": false,
    "validatingHost": false,
-   "validatingMacaroon": false,
+   "validatingMacaroon": true,
  }
`;

exports[`reducers onboardingReducer should handle initial state 1`] = `
Object {
  "alias": "",
  "autopilot": false,
  "chain": "bitcoin",
  "connectionCert": "",
  "connectionHost": "",
  "connectionMacaroon": "",
  "connectionString": "",
  "connectionType": "create",
  "isSkipBackupDialogOpen": false,
  "lndConnect": null,
  "name": "",
  "network": "mainnet",
  "passphrase": "",
  "password": "",
  "seed": Array [],
  "validatingCert": false,
  "validatingHost": false,
  "validatingMacaroon": false,
}
`;
